tests/test_FBbasis2D.py

```diff
@@ -66,18 +67,20 @@ def testFBBasis2DEvaluate(self):
                  2.37066082e-02,   4.88805735e-03,   1.47870707e-03,   7.63376018e-03,
                 -5.60619559e-03,   1.05165081e-02,   3.30510143e-03,  -3.48652120e-03,
                 -4.23228797e-04,   1.40484061e-02
-              ]
+              ],
+            dtype=self.dtype
         )
         result = self.basis.evaluate(coeffs)
 
         self.assertTrue(np.allclose(
             result,
-            np.load(os.path.join(DATA_DIR, 'fbbasis_evaluation_8_8.npy')).T  #RCOPT
+            np.load(os.path.join(DATA_DIR, 'fbbasis_evaluation_8_8.npy')).T,  #RCOPT
+            atol=1e-6 if self.dtype == np.float32 else 1e-8
```

Comment:
There's no way this is handled elegantly when both arguments to `allclose` are `float32`? If not, we should probably hardcode these tolerances in one place and call something like `test_tolerance(dtype)` when needed.

Reply:
1) Negative, their default 1e-8 is already meant to be flexible for singles, and for many things it is. I think if you expect to feed certain solvers noise this expectation will not hold up.

2) I can do that but they're not all the same (well I can make them, but I would prefer them tighter). We'd have to select the crudest value (so probably 1e-5) or go back to adhocs.

Reply:
I have implemented `utest_tolerance` at 1e-5 and replaced the inline `atol` statements.

Reply:
If I understand correctly, `np.isclose` has two tolerances: relative (default `1e-5`) and absolute (default `1e-8`). The latter is only relevant for values close to zero (where a relative error doesn't make much sense, since it would be `1e-5` times zero). In other words, the absolute tolerance is there to make sure that machine epsilon and zero are considered “close.” Yet machine epsilon for `float32` is bigger than `1e-8` (`np.finfo(np.float32).eps` gives `1.192e-7` for me), so it's not going to catch those. For example, the following fails on my laptop:
```
x = np.array(1 / 3, dtype=np.float32) * 3 - np.array(1, dtype=np.float32)
np.isclose(x, 0)
```
(I get `x = 2.980e-08`.) Looking at the reference data in this particular test, we have values like `1e-17`, which I'm guessing is the cause for the test failure.

Long story short, I think we can feel safe setting our absolute tolerance a few orders of magnitude higher than machine epsilon. It's the relative tolerance that's important (and that stays at `1e-5`). For the exceptional cases where we have values that are essentially zero, it's ok to be a little loose.

(The flip side of this is that we can probably tighten the tolerances for `float64`. A relative error of `1e-5` and an absolute error of `1e-8` is pretty high for double precision. But that's a separate issue.)

Reply:
I agree about their atol and rtol. Regarding that example, you've made my point about being more careful about mixing types. `x` is a double; the arithmetic is mixed mode.  Been down that road many times, sometimes I wonder if too many :D ...

```
>>> x = np.array(1 / 3, dtype=np.float32) * 3 - np.array(1, dtype=np.float32); np.isclose(x, np.float32(0))
False
>>> x = np.array(1 / 3, dtype=np.float32) * np.float32(3) - np.array(1, dtype=np.float32); np.isclose(x, np.float32(0))
True
```
alternately
```
>>> x = np.zeros(2, dtype=np.float32)
>>> x[0] = 1 / 3
>>> x[1] = 1
>>> x[0] *= 3
>>> x[1] - x[0]
0.0
>>> np.isclose(x[1] - x[0], 0)
True
>>> (x[1] - x[0]).dtype
dtype('float32')
>>> float(x[1] - x[0]).hex()
'0x0.0p+0'
>>> ui32 = (x[1] - x[0]).view(dtype=np.uint32)
>>> ui32
0
```

I feel reasonably good about numpy's expectations in their domain, but what ASPIRE does is a little more involved (multi method) right.  Unlike some other fields, I don't have a great pulse on what the propagation of uncertainty looks like numerically in your field yet (though I still have my intuitions)... if you are comfortable with the tolerances I'm satisfied.  One sneaking intuition is that if we stay vigilant some of these can be improved, but maybe not today.

Reply:
Huh… I didn't realize it's quite that bad. (So it's good that we're getting away from this stuff.) I'm trying to understand what's happening in that particular example for things to come out to exactly zero, but my familiarity with IEEE 754 arithmetic is not too good… Perhaps we can discuss at some point. My point is there's no reason for errors due to non-associativity to be smaller than `1e-8` in single precision, so it seems odd to have that as a default absolute tolerance.

At any rate, this is fine for now